### PR TITLE
Obromios bug fixes

### PR DESCRIPTION
**Pull request by obromios**
This is similar to PR #49 but with experimental features and calibration enhancements removed.

## Fixes for issue #33 (Intermittent errors in reading mag data)
- Add a while loop in read_mag() to check ST1 is set correctly
- Added a time out for the while loop
- Added a setting, mpu.set_mag_time_out to control length of timeout

## Fix for issue #34 (Gyro readings noisy)
- Inverted application of setting.gyro_fchoice in initMPU9250()

## Fix for issue #39 (initMPU9250() is not initializing the accelerometer)
- Added a reset at beginning of initMPU9250()

## Fix for issue #45 (Consistent use of fuse values)
- In collect_mag_data(), used mag_bias_factory[i] as a multiplicative factor in calculation of scale[i]

## Fix for issues #46, #48, #42 and #30.

These issues are 

- What is the correct way to transform the raw readings for the quaternion update (#46)
- Using setMagneticDeclination does not work properly (#48)
- Yaw offset problem (#42)
- Wrong Linear Acceleration readings (#30)
- Root cause of all these issues was related to the choice of coordinate systems.
- The fix was to provide a setting to switch between coordinate systems.
- At present, two options are provided, NED and NWU.  NWU is only partially supported (see below)

## Partial support for NWU coordinate systems
- The commit on January 5 (455a9efd109cb5c9f0bc96710979cb451ef2fae9), made a change to the coordinate system being fed into the quaternion filter (see issue #46).  Prior to this, the coordinate system was the same used by Kris Winer, NED. After this change there was a new coordinate system, which until clarification is being called NWU.
- This was a breaking change that caused a number of other issues (#48, #42, and #30). There maybe other problems caused by the breaking change that have yet to be detected.   Accordingly, a setting has been introduced that allows the coordinate system to be set back to NED.  This fixes these issues.
- There is also a settings option to change the coordinate system to NWU, but using this option will cause the issues in #48, #42, and #30 to reappear.  As well, there is an extremely long settling time for rotations in the horizontal plane.  It may possible to rectify each issue (and other possible undetected issues) by making changes specific the NWU coordinate system.
- Given using the NWU system currently causes breaking changes, it is recommended that the default coordinate system be reverted to NED, until the NWU coordinate option is fully implemented. If no further work is to be done on the NWU coordinate system, it is recommended that it is removed as an option.

### CODE REVIEW

File: MPU9250.h [hunk 1]
```diff
@@ -349,12 +370,13 @@ class MPU9250_ {
...
         uint8_t c = read_byte(mpu_i2c_addr, GYRO_CONFIG);  // get current GYRO_CONFIG register value
-        c = c & ~0xE0;                                     // Clear self-test bits [7:5]
-        c = c & ~0x03;                                     // Clear Fchoice bits [1:0]
-        c = c & ~0x18;                                     // Clear GYRO_FS_SEL bits [4:3]
-        c = c | (uint8_t(setting.gyro_fs_sel) << 3);       // Set full scale range for the gyro
-        c = c | (uint8_t(setting.gyro_fchoice) & 0x03);    // Set Fchoice for the gyro
+        c = c & ~0xE0;                                        // Clear self-test bits [7:5]
+        c = c & ~0x03;                                        // Clear Fchoice bits [1:0]
+        c = c & ~0x18;                                        // Clear GYRO_FS_SEL bits [4:3]
+        c = c | (uint8_t(setting.gyro_fs_sel) << 3);          // Set full scale range for the gyro
```
Comment (hideakitai) [inline MPU9250.h:378]: No need to change these lines

File: MPU9250.h [hunk 2]
```diff
@@ -106,8 +112,9 @@ class MPU9250_ {
 
     // Other settings
     bool has_connected {false};
-    bool b_ahrs {true};
-    bool b_verbose {false};
+    bool b_ahrs{true};
+    bool b_verbose{false};
```
Comment (hideakitai) [inline MPU9250.h:116]: No need to delete spaces

File: MPU9250.h [hunk 3]
```diff
@@ -106,8 +112,9 @@ class MPU9250_ {
...
+    int mag_time_out {1000};  // Timeout for read_mag in ms
```
Comment (hideakitai) [inline MPU9250.h:117]: No need to add timeout. Please see 2a151e0eec30f26838af99829f634d260e4b6e21

File: MPU9250.h [hunk 4]
```diff
@@ -157,6 +164,11 @@ class MPU9250_ {
         b_ahrs = b;
     }
 
+    void set_mag_time_out(const int milliseconds){
+        mag_time_out = milliseconds;
+    }
+    
+
```
Comment (hideakitai) [inline MPU9250.h:171]: No need to add this function. Please see 2a151e0eec30f26838af99829f634d260e4b6e21

File: MPU9250.h [hunk 5]
```diff
@@ -349,12 +370,13 @@ class MPU9250_ {
 
         // Set gyroscope full scale range
         // Range selects FS_SEL and GFS_SEL are 0 - 3, so 2-bit values are left-shifted into positions 4:3
+
```
Comment (hideakitai) [inline MPU9250.h:373]: No need to add

File: MPU9250.h [hunk 6]
```diff
@@ -349,12 +370,13 @@ class MPU9250_ {
...
-        c = c & ~0xE0;                                     // Clear self-test bits [7:5]
-        c = c & ~0x03;                                     // Clear Fchoice bits [1:0]
-        c = c & ~0x18;                                     // Clear GYRO_FS_SEL bits [4:3]
-        c = c | (uint8_t(setting.gyro_fs_sel) << 3);       // Set full scale range for the gyro
-        c = c | (uint8_t(setting.gyro_fchoice) & 0x03);    // Set Fchoice for the gyro
+        c = c & ~0xE0;                                        // Clear self-test bits [7:5]
+        c = c & ~0x03;                                        // Clear Fchoice bits [1:0]
+        c = c & ~0x18;                                        // Clear GYRO_FS_SEL bits [4:3]
+        c = c | (uint8_t(setting.gyro_fs_sel) << 3);          // Set full scale range for the gyro
+        c = c | (uint8_t(~setting.gyro_fchoice) & 0x03);       // Set Fchoice for the gyro by loading complement of Fchoice into Fchoice_b bits.
```
Comment (hideakitai) [inline MPU9250.h:379]: Ok, please see 4a58e270562dc4eab219e81ad0d510a0f4725afb

File: MPU9250.h [hunk 7]
```diff
@@ -396,6 +418,7 @@ class MPU9250_ {
         mag_bias_factory[0] = (float)(raw_data[0] - 128) / 256. + 1.;  // Return x-axis sensitivity adjustment values, etc.
         mag_bias_factory[1] = (float)(raw_data[1] - 128) / 256. + 1.;
         mag_bias_factory[2] = (float)(raw_data[2] - 128) / 256. + 1.;
+
```
Comment (hideakitai) [inline MPU9250.h:421]: No need to add line

File: MPU9250.h [hunk 8]
```diff
@@ -494,6 +517,14 @@ class MPU9250_ {
...
     void read_mag(int16_t* destination) {
         uint8_t raw_data[7];                                                 // x/y/z gyro register data, ST2 register stored here, must read ST2 at end of data acquisition
+        uint8_t ST1;
+        long int wait_time;
+        long int start_mag = millis();
+        do {
+            ST1 = read_byte(AK8963_ADDRESS, AK8963_ST1);
+            wait_time = millis() - start_mag;
+        } while(!(ST1 & 0x01) && (wait_time < mag_time_out));
+        if ((!ST1 & 0x01) && b_verbose) Serial.println("Warning - read_mag() timed out");
```
Comment (hideakitai) [inline MPU9250.h:527]: No need to add these lines. Please see 2a151e0eec30f26838af99829f634d260e4b6e21

File: MPU9250.h [hunk 9]
```diff
@@ -668,7 +699,9 @@ class MPU9250_ {
         MAG_OUTPUT_BITS mag_output_bits_cache = setting.mag_output_bits;
         setting.mag_output_bits = MAG_OUTPUT_BITS::M16BITS;
         initAK8963();
+
         collect_mag_data_to(mag_bias, mag_scale);
+
```
Comment (hideakitai) [inline MPU9250.h:704]: No need to add these lines

File: MPU9250.h [hunk 10]
```diff
@@ -325,6 +343,9 @@ class MPU9250_ {
         gyro_resolution = get_gyro_resolution(setting.gyro_fs_sel);
         mag_resolution = get_mag_resolution(setting.mag_output_bits);
 
+        
+        write_byte(mpu_i2c_addr, PWR_MGMT_1, 0x80);  // Write a one to bit 7 reset bit; toggle reset device
+        delay(100);
```
Comment (hideakitai) [inline MPU9250.h:348]: Ok, please see ebe2a8830baafddacfb1980fba38fa8f4fa1ad6c

**Review by hideakitai — CHANGES_REQUESTED** [review]
Sorry for the late reply. I've checked your code and there are many unnecessary changes so please revert them.

About #33, your suggestion is not correct and I've made changes in 2a151e0eec30f26838af99829f634d260e4b6e21. 

About fixes for #34 and #39 are OK but I can't merge them because this pull request is not small to merge right now... 

I've made example commits about #34 and #39 4a58e270562dc4eab219e81ad0d510a0f4725afb and ebe2a8830baafddacfb1980fba38fa8f4fa1ad6c, so if you are ok, I can merge them by myself. If you can create small pull requests each for them, I will wait for it.

The rest of the pull requests will be reviewed next week... Thank you for your contribution!

**Comment by hideakitai** [comment]
@obromios I've also fixed coordinate issues in 23fb7d961164543e55c2815e0aad5f378aa672a3 .

I've selected to revert calculation to NED and just flip the direction to get euler angles. Maybe a series of commits fixes all issues you have listed, and if you agree and can confirm if it works, I will merge the branch. How do you think about it?

**Comment by obromios** [comment]
@hideakitai, I have checked and the NED angles appears to be correct.  I think all the other changes appear to work. It is looking good.

Chris

**Comment by hideakitai** [comment]
@obromios If you can agree with my changes, can I merge the branch above into the main branch instead of this pull request?

**Comment by obromios** [comment]
I assume by the branch above you mean https://github.com/hideakitai/MPU9250/tree/bugfix_obromious.  I agree  with merging that branch.

**Comment by hideakitai** [comment]
@obromios Thank you, please open the new pull request if you have additional suggestions. Thank you so much for your huge contribution!

**Comment by hideakitai** [comment]
Fixed in v0.4.3

**Pull request closed without merging**